target/linux/ath79/dts/ar9342_pcs_cr5000.dts

```diff
@@ -138,11 +138,7 @@
 	status = "okay";
 
 	ath9k: wifi@0,0 {
-		compatible = "pci168c,0030";
```

Comment:
Please keep the compatible and reg properties around. Reason being: The AR9300/AR9380 chip on the pcie is real hardware. The unit-address 0,0 in wifi@0,0 is supposed to have a corresponding reg property, otherwise the dtc compiler will complain about it.

Reply:
Hmmm...it looks like this has been a rabbit hole.  I'm not sure there are actually any changes to make to the DT after all.  However 5GHz wifi without the changes is throwing errors that don't show up with ar71xx -- I thought perhaps the issue was a missing ap94_pci_init style pci fixup, and was trying to borrow from the few boards that also use that function.  I guess I'll have to get some logs and post to the mailing list as I have no idea what else to try.

Reply:
What ath9k errors are you seeing on the 5GHz wifi? Do you get something like "-5 failed to initialize" or do the errors pop up during use? Also, what does the caldata on 0x5000 look like? Is it just the valid mac at the 0x2 offset or does it look like it's complete? If the caldata at 0x5000 looks complete you could try to load into the chip and see if it fixes the problem reliably .

dts changes:

         ath9k: wifi@0,0 {
              compatible = "pci168c,0030";
              reg = <0x0000 0 0 0 0>;
              qca,no-eeprom; /* instructs ath9k to not try the eeprom */
          };

 ath79/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom

          ....
         "ath9k-eeprom-pci-0000:00:00.0.bin")
         case $board in
             ...
             pcs,cr5000)
                    ath9k_eeprom_extract "art" 20480 4096
                    ;;
             ....

(The 4096 size might be a little big.)

Reply:
Errors during use and there is no caldata (all 0x00 except the mac address at 0x02 offset into what would be 5ghz caldata (0x5000)).  I'm preparing a message for the mailing list and will post the link here.

Reply:
>root@OpenWrt:/# iwinfo wlan1 scan
>[   36.472062] ath: phy1: Unable to reset hardware; reset status -5 (freq 5180 MHz)
>[   36.710068] ath: phy1: Unable to reset channel, reset status -5
>[   36.716420] IPv6: ADDRCONF(NETDEV_UP): wlan1: link is not ready
>[   36.950537] ath: phy1: Unable to reset channel, reset status -5
>[repeats]
>[   39.895120] ath: phy1: Could not kill baseband RX
>[repeat the above two error messages]

Ok, so the good news is that the hardware initializes successfully with the eeprom.
Do you think you can dump its content/hexdump somewhere onto the net (github gist/pastebin)? The data is available through debugfs 
 /sys/kernel/debug/ieee80211/phy0/ath9k/eeprom 
(base_eeprom is interesting too.)

Also, (but this is a moonshot). Can you try enabling
CONFIG_ATH9K_SUPPORT_PCOEM (Support chips used in PC OEM cards)

option for the kmod-ath9k package?

I'm asking because it's not enabled by default to safe space and I think it can affect the ath9k's AR9300 calibration.

Reply:
Here's a gist with the eeprom and base_eeprom hexdumps: https://gist.github.com/cshoredaniel/45a217fe76cc98f836fca1ebb06a6258.
Now to try PCOEM support.

Reply:
So PCOEM fixes the MAC address issue, but not the error messages...but I think that may be a function of wrong wifi@0,0 and reg ... I suspect those are funky, based on what I've seen on other ap94_pci_init devices and their DTS.

Reply:
Interestingly if I use reg = <0x9000 0 0 0 0> (borrowed from dir-825-b1):
I get the following eeprom dumps: https://gist.github.com/cshoredaniel/4a42897385e3ed90d9561e11b91db9f1.  The eeprom.txt is the same but base_eeprom.txt is now matches the actual hardware in terms of advertised capabilities etc.  Any thoughts?

NB: This is a ca. 2012-2013 device so I think the quoted standards might not be accurate for these devices.

Reply:
@mkresin @nbd @blogc I wonder if the issue is 'big-endian 0' -- I recall seeing on another device / thread some discussion of mixed up ath9k endianness issues, but can't find it at the moment.

Reply:
Ok, hexdump was not the smart move. But I think it got the result I was after.


> I wonder if the issue is 'big-endian 0' -- I recall seeing on another device / thread some discussion of mixed up ath9k endianness issues, but can't find it at the moment.

The big-endian flag is derived from the eeprom content. So, if the eeprom byte-order is wrong the flag value can get messed up as well.

Can you put this patch into /package/kernel/mac80211/patches/ath/999-no-eep-swap.patch


    --- a/drivers/net/wireless/ath/ath9k/init.c	2018-12-12 21:47:44.279524942 +0100
    +++ b/drivers/net/wireless/ath/ath9k/init.c	2018-12-12 22:02:42.555742113 +0100
    @@ -649,7 +649,6 @@ static int ath9k_of_init(struct ath_soft
     		ether_addr_copy(common->macaddr, mac);
     
     	ah->ah_flags &= ~AH_USE_EEPROM;
    -	ah->ah_flags |= AH_NO_EEP_SWAP;
     
     	return 0;
     }

recompile (preferably with reg = <0 0 0 0>; and no disable-2ghz in the dts ) ?

(Yes: masking AH_USE_EEPROM seems also be wrong, because you have an eeprom (or otp?). But one step at a time)

Reply:
@chunkeey Testing, and thank you very much for all your help!

Reply:
Ok, mac address is now correct, but still get the error messages when scanning (I presume that is to be expected because of AH_USE_EEPROM masking).
What's next?

Reply:
> Ok, so also removing masking AH_USE_EEPROM (along with the previous change) gets wifi working right, with no dts changes.

Ok, this is what we are after. 

I think we can now also fix the ar71xx target too.

    --- a/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c	2018-12-14 16:25:10.400561574 +0100
    +++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c	2018-12-14 16:27:05.627316026 +0100
    @@ -136,6 +136,7 @@ static struct mdio_board_info cr5000_mdi
     static void __init cr5000_setup(void)
     {
     	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
    +	struct ath9k_platform_data *pdata;
     
     	ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);
     	gpio_request_one(CR5000_GPIO_LED_POWER_ENABLE,
    @@ -154,8 +155,11 @@ static void __init cr5000_setup(void)
     	ath79_register_usb();
     	ath79_register_wmac(art + CR5000_WMAC_CALDATA_OFFSET, art + CR5000_WMAC_MAC_OFFSET);
     	ap94_pci_init(NULL, NULL, NULL, art + CR5000_PCIE_WMAC_OFFSET);
    +	pdata = ap9x_pci_get_wmac_data(0);
    +	if (pdata)
    +		pdata->use_eeprom = true;
     
    -        ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
    +	ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
     
     	ath79_register_mdio(0, 0x0);
     

Can you please test if this patch produces the desired outcome on the ar71xx?

Reply:
> Question is what's the right way to make that patch a DTS option (since I presume normally those flags are correct).

I guess we could ask @xdarklight too. Martin: Daniel reported an issue that he currently is facing with the integrated AR9380 pcie on his CR5000 device with the new device-tree enabled ath79 target. All the information is in this thread and the problem has been debugged to the clearing of the AH_USE_EEPROM flag in ath9k_of_init.

https://elixir.bootlin.com/linux/v4.20-rc6/source/drivers/net/wireless/ath/ath9k/init.c#L645
https://elixir.bootlin.com/linux/v4.20-rc6/source/drivers/net/wireless/ath/ath9k/init.c#L646
 
In my opinion, the AH_USE_EEPROM should only be cleared when the "qca,no-eeprom" property is present. I have something like this in mind:

    diff --git a/drivers/net/wireless/ath/ath9k/init.c b/drivers/net/wireless/ath/ath9k/init.c
    --- a/drivers/net/wireless/ath/ath9k/init.c
    +++ b/drivers/net/wireless/ath/ath9k/init.c
    @@ -636,14 +636,16 @@ static int ath9k_of_init(struct ath_softc *sc)
     		ret = ath9k_eeprom_request(sc, eeprom_name);
     		if (ret)
     			return ret;
    +
    +		ah->ah_flags &= ~AH_USE_EEPROM;
     	}
     
     	mac = of_get_mac_address(np);
     	if (mac)
     		ether_addr_copy(common->macaddr, mac);
     
    -	ah->ah_flags &= ~AH_USE_EEPROM;
    -	ah->ah_flags |= AH_NO_EEP_SWAP;
    +	if (of_property_read_bool(np, "qca,caldata-endianness-quirk"))
    +		ah->ah_flags |= AH_NO_EEP_SWAP;
     
     	return 0;
     }


What do you think? Or do you know of a case/device where this would fail?

Reply:
> > Ok, so also removing masking AH_USE_EEPROM (along with the previous change) gets wifi working right, with no dts changes.
> 
> Ok, this is what we are after.
> 
> I think we can now also fix the ar71xx target too.
> 
> ```
> --- a/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c	2018-12-14 16:25:10.400561574 +0100
> +++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c	2018-12-14 16:27:05.627316026 +0100
> @@ -136,6 +136,7 @@ static struct mdio_board_info cr5000_mdi
>  static void __init cr5000_setup(void)
>  {
>  	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
> +	struct ath9k_platform_data *pdata;
>  
>  	ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);
>  	gpio_request_one(CR5000_GPIO_LED_POWER_ENABLE,
> @@ -154,8 +155,11 @@ static void __init cr5000_setup(void)
>  	ath79_register_usb();
>  	ath79_register_wmac(art + CR5000_WMAC_CALDATA_OFFSET, art + CR5000_WMAC_MAC_OFFSET);
>  	ap94_pci_init(NULL, NULL, NULL, art + CR5000_PCIE_WMAC_OFFSET);
> +	pdata = ap9x_pci_get_wmac_data(0);
> +	if (pdata)
> +		pdata->use_eeprom = true;
>  
> -        ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
> +	ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
>  
>  	ath79_register_mdio(0, 0x0);
> ```
> 
> Can you please test if this patch produces the desired outcome on the ar71xx?

Sorry just noticed this...will test ar71xx if not tonight then tomorrow.

Reply:
> ok the ar71xx patch didn't have the desired effect (wrong mac address), HOWEVER I am going to try changing pdata = ap9x_pci_get_wmac_data(0); to pdata = ap9x_pci_get_wmac_data(1); as the pcie 5ghz radio is not only the second radio registered in general, but the pci card has two radios, but the 2ghz radio is not used on this device.

Oh heh. I think I know what's going on now. 

The AR1022 / AR9342 has a builtin (in the SoC) ath9k on the AHB, right? And the AR9380 chip is connected over the SoCs PCIe link. This makes the 5GHz AR9380 the **first** and only PCIe device. 
Don't get confused by the terminology. The [AR9380](https://wikidevi.com/files/Atheros/specsheets/AR9380.pdf) has a dual band radio, but it only has one MAC and one host interface (pcie) .

So the 

     ap94_pci_init(NULL, NULL, NULL, art + CR5000_PCIE_WMAC_OFFSET);

should have been:

     ap94_pci_init(NULL, art + CR5000_PCIE_WMAC_OFFSET, NULL, NULL);

The ath79_register_wmac() and ap94_pci_init() each serves a different purpose since the
integrated wifi in the AR1022/AR9342 SoC is independent of the AR9380. 

---
    --- a/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c
    +++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c
    @@ -136,6 +136,7 @@ static struct mdio_board_info cr5000_mdio0_info[] = {
     static void __init cr5000_setup(void)
     {
     	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
    +	struct ath9k_platform_data *pdata;
     
     	ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);
     	gpio_request_one(CR5000_GPIO_LED_POWER_ENABLE,
    @@ -153,9 +154,12 @@ static void __init cr5000_setup(void)
     					cr5000_gpio_keys);
     	ath79_register_usb();
     	ath79_register_wmac(art + CR5000_WMAC_CALDATA_OFFSET, art + CR5000_WMAC_MAC_OFFSET);
    -	ap94_pci_init(NULL, NULL, NULL, art + CR5000_PCIE_WMAC_OFFSET);
    +	ap94_pci_init(NULL, art + CR5000_PCIE_WMAC_OFFSET, NULL, NULL);
    +	pdata = ap9x_pci_get_wmac_data(0);
    +	if (pdata)
    +		pdata->use_eeprom = true;
     
    -        ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
    +	ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
     
     	ath79_register_mdio(0, 0x0);

Reply:
Doh!  Actually the following is what's needed:

```
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c
index 0851899b40..7c41e7fbc0 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-cr5000.c
@@ -136,6 +136,7 @@ static struct mdio_board_info cr5000_mdio0_info[] = {
 static void __init cr5000_setup(void)
 {
        u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
+       struct ath9k_platform_data *pdata;
 
        ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);
        gpio_request_one(CR5000_GPIO_LED_POWER_ENABLE,
@@ -153,10 +154,12 @@ static void __init cr5000_setup(void)
                                        cr5000_gpio_keys);
        ath79_register_usb();
        ath79_register_wmac(art + CR5000_WMAC_CALDATA_OFFSET, art + CR5000_WMAC_MAC_OFFSET);
-       ap94_pci_init(NULL, NULL, NULL, art + CR5000_PCIE_WMAC_OFFSET);
-
-        ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
+       ap91_pci_init(NULL, art + CR5000_PCIE_WMAC_OFFSET);
+       pdata = ap9x_pci_get_wmac_data(0);
+       if (pdata)
+               pdata->use_eeprom = true;
 
+       ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_RGMII_GMAC0);
        ath79_register_mdio(0, 0x0);
 
        ath79_init_mac(ath79_eth0_data.mac_addr, art + CR5000_MAC0_OFFSET, 0);
```
Note that ap94_pci_init was actually wrong all along (wrong function numbers), and ap91_pci_init is the right function since it's a single dual band radio not two single (or dual) band radios on the pcie.

Reply:
Also note this is with the ath9k patch above still applied, which probably is relevant.

Reply:
> Ok, mac address is now correct, but still get the error messages when scanning (I presume that is to be expected because of AH_USE_EEPROM masking).
> What's next?

I think that may have been because of PCOEM setting that I had tried.  Going try again with working EEPROM situation but no endianness quirk setting (and PCOEM has been disabled for ar71xx tests and USE_EEPROM tests already and I confirmed is disabled now).  Or is it that I'm reading the log on the quirk wrong and when the qca,caldata-endianness-quirk it does NOT add the AH_NO_EEP_SWAP flag? (in which case the test I'm doing now should not work).

Reply:
ok, I'm going to try with a distclean to verify the endianness quirk is not necessary.  Latest test succeeded.  It looks like the real issue was the masking out of USE_EEPROM.